### ISSUE TITLE
`ghcup provision` (or `ghcup setup-env`) etc.

### ISSUE THREAD

**Issue by hasufell** [haskell/ghcup-hs#989]
To discuss: instead of running `ghcup install ghc 9.2.8 && ghcup install ghc 9.4.8 && ...` one could imagine a way to declaratively specify the whole toolchain setting and let ghcup execute that description as a whole, e.g.

```
GHC: [9.2.8, 9.4.8*, 9.6.3]
Stack: [latest*]
Cabal: [latest*]
HLS: [recommended*]
```

or some-such. There's some design-space to explore (e.g. where to read the data/file from), whether to support some form of `.ghcup` file in PWD, etc.

**Comment by runeksvendsen** [comment]
As discussed privately with @hasufell: note that whatever DSL the above example is parsed into would also be useful for https://github.com/haskell/ghcup-hs/issues/64. In the integration test, we could imagine a test case that installs the above tools/versions and asserts something about the resulting environment, and in that context it would be useful to have a DSL for expressing a set of tools to install/set.

**Comment by runeksvendsen** [comment]
One could also imagine a new command `ghcup shell` which drops the user into a shell in which the tools specified by `.ghcup` are available (ie. instead of installing them globally).

**Comment by hasufell** [comment]
> Probably we are seeing things different or I most likely am wrongly assuming some bits but, isn't latest also stateful?

No, it's not stateful in terms of the filesystem. It's hardcoded in the ghcup metadata file. You can pin the ghcup metadata file if you want perfect reproducibility: `ghcup -s https://raw.githubusercontent.com/haskell/ghcup-metadata/9ce5a2cf018443f8ea9c9f4a06726279066e05e9/ghcup-0.0.9.yaml tui`

I just think you're breaking the "reproducible provisioning" if you allow `latest-installed`. Now it's not reproducible anymore and what's the point?

> I see how latest-installed is stateful. But it is the current semantic of vscode. If we agree to move the logic resolution out of vscode I think this is needed.

Just because VSCode does something doesn't mean I want that logic in GHCup. VSCode as a whole has a very different UX than, say, vim. GHCup is meant to follow the unix principles... VSCode doesn't really. So we can do more outrageous things in VSCode.

The `latest-installed` in VSCode is probably due to the fact that we don't want to constantly install new things when the editor boots up. This is very different from the use case of a user saying `ghcup provision` to get a reproducible environment.
